ibtk/include/ibtk/MarkerPoints.h

```diff
@@ -0,0 +1,190 @@
+     */
+    bool contains(const IBTK::Point& position) const;
+
+    /**
+     * Remove all marker points which do not lie in cells uniquely owned by
+     * this patch.
+     */
+    std::tuple<std::vector<int>, EigenAlignedVector<IBTK::Point>, EigenAlignedVector<IBTK::Vector> > prune();
+
+    /*!
+     * Return the @p index-th marker point stored by the present Patch.
+     */
+    std::tuple<int, IBTK::Point, IBTK::Vector> operator[](const unsigned int index) const;
+
+    /*!
+     * Return the number of marker points presently stored by this object.
+     *
+     * @note This may contain points outside the cells uniquely owned by this
+     * Patch if prune() is not called first.
+     */
+    std::size_t size() const;
+
+private:
+    // Marker-point data. Stored in a packed format so we can directly call
+    // Fortran routines on it.
+    std::vector<int> d_indices;
+    std::vector<double> d_positions;
+    std::vector<double> d_velocities;
+
+    // Box for the patch itself.
+    SAMRAI::hier::Box<NDIM> d_patch_box;
+
+    // The unique subset of index space assigned to this Patch.
+    std::vector<SAMRAI::hier::Box<NDIM> > d_nonoverlapping_patch_boxes;
+
+    // Data for computing cell indices. Passed along to IndexUtilities::getCellIndex();
+    SAMRAI::hier::Box<NDIM> d_domain_box;
+    std::array<double, NDIM> d_x_lo;
+    std::array<double, NDIM> d_x_up;
+    std::array<double, NDIM> d_dx;
+
+    // Allow MarkerPoints to directly manipulate the arrays owned by this
+    // class for timestepping.
+    friend class MarkerPoints;
+};
+
+/*!
+ * Implementation of marker points, also sometimes called particles. These
+ * points are akin to IB points except they are simply advected by the fluid
+ * and do not apply any force upon it.
+ */
+class MarkerPoints : public SAMRAI::tbox::DescribedClass
+{
+public:
+    /**
+     * Constructor.
+     */
+    MarkerPoints(const std::string& name,
+                 SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                 const EigenAlignedVector<IBTK::Point>& positions,
+                 const EigenAlignedVector<IBTK::Point>& velocities);
+
+    void reinit(const EigenAlignedVector<IBTK::Point>& positions, const EigenAlignedVector<IBTK::Point>& velocities);
+
+    /**
+     * Get the MarkerPatch associated with the present level and local patch number.
+     */
+    const MarkerPatch& getMarkerPatch(const int ln, const int local_patch_num) const;
+
+    /**
+     * Get the total number of marker points stored across all processors.
+     */
+    std::size_t getNumberOfMarkers() const;
+
+#if 0
+    /**
+     * Save the present state of the object to a SAMRAI database.
+     */
+    void putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db) override;
+#endif
+
+    /**
+     * Set marker velocities with some given velocity field @p u_idx and
+     * interaction kernel @p kernel.
```

Comment:
Do we list the available kernels somewhere? Or maybe point to `LEInteractor` for kernel names?